FunctionDepsFinder: Check for declarations inside another declaration

marcosps opened the pull request:
Closes: #44

giulianobelinassi commented on libcextract/FunctionDepsFinder.hh:
```diff
@@ -43,6 +43,33 @@ class ClosureSet
+   * Check if the same Decl was already added into the closure but for a
+   * different typename.
+   */
+  TypedefDecl *insideRangeOfDecl(TypedefDecl *decl)
```
Please use `Decl` and not only `TypedefDecl`.

marcosps replied:
There were some complaints from the compiler when I was using Decl, so I used the more specialized type. I'll check it tomorrow.

giulianobelinassi commented on libcextract/FunctionDepsFinder.hh:
```diff
@@ -43,6 +43,33 @@ class ClosureSet
+  {
+    SourceRange range = decl->getSourceRange();
+
+    for (auto it = Dependencies.begin(); it != Dependencies.end(); ++it) {
```
This algorithm is O(_n_ ²). If the closure is small, I think this is acceptable, but if we want to do this extensively we may be in a better shape to sort by SourceLocation something to reduce it to O(n log n).

marcosps replied:
The idea in this case was to submit the solution to check with our CI. Also, for my current LP, it worked.

I'll think how this could be improved.

giulianobelinassi commented on libcextract/FunctionDepsFinder.hh:
```diff
@@ -43,6 +43,33 @@ class ClosureSet
+        if (cur_range == range)
+          continue;
+
+        if (cur_range.getBegin() == range.getBegin() &&
```
You can use Contains_From_LineCol and get the Decl with larger Range.

marcosps replied:
Cool, I'll use it

giulianobelinassi commented on libcextract/FunctionDepsFinder.hh:
```diff
@@ -43,6 +43,33 @@ class ClosureSet
+    SourceRange range = decl->getSourceRange();
+
+    for (auto it = Dependencies.begin(); it != Dependencies.end(); ++it) {
+      if (TypedefDecl *cur_decl = dynamic_cast<TypedefDecl *>(*it)) {
```
Please avoid using dynamic_cast<type*> and use dyn_cast. This works on clang builds with RTTI disabled.

marcosps replied:
I just copied this line from the FunctionDepsFinder :P

Thanks for the suggestion

marcosps commented:
So I addressed all your comments besides the algorithm. Also, I added a testcase, which was simple to craft. Tomorrow I plans to rethink the algorithm so we can have this merged.
